repair/repair_task.hh

```diff
@@ -82,6 +89,10 @@ class repair_module : public task_manager::module {
     //
     // The order of shutting down. Can we use common gate? Can we get rid of shutdown methods?
 private:
+    // Used to allow shutting down repairs in progress, and waiting for them.
+    seastar::gate _repair_gate;
```

Comment:
why not use the base class `_gate`?

Reply:
I haven't done that because I did not understand why repair shutdown is divided into three part and I didn't wanted to mess it up. I think I finally managed to get it but I need your confirmation.

Repair shutdown is divided into 3 parts (`main.cc`):
- `repair_service::stop()`
- repair api is unregistered
- `repair_service::shutdown()`
which happen to be called bottom to top.

This is done because we want to stop all the repairs, while api is still available. Then, unregister the api. And then, close whole `repair_service`.

If this is true, then I should surely move `module::stop()` from `repair_service::stop()` to `repair_service::shutdown()`.
And then we could freely use the `_gate` and maybe even delete `with_gate` from:
https://github.com/scylladb/scylladb/blob/6b180bf2c7830a5930ffcc87a9e4cf3a25db9092/repair/repair.cc#L474-L485
Since it is already protected by `task_manager::task` gate.

One thing that comes to mind here is if task manager api should have been unregistered. Some of the already existing services are unregistered, some are not. I do not see the pattern of choice.

Also I think that the shutdown should happen in the following order:
- stop tasks
- unregister services' apis
- (unregister task manager api)
- close task_manager::modules
- close task_manager
- close services
Thus, be called in `main` in reversed order.

What do you think?

Reply:
> I haven't done that because I did not understand why repair shutdown is divided into three parts and I didn't wanted to mess it up. I think I finally managed to get it but I need your confirmation.
> 
> Repair shutdown is divided into 3 parts (`main.cc`):
> 
> * `repair_service::stop()`
> * repair api is unregistered
> * `repair_service::shutdown()`
>   which happen to be called bottom to top.
> 
> This is done because we want to stop all the repairs, while api is still available. Then, unregister the api. And then, close whole `repair_service`.
> 
> If this is true, then I should surely move `module::stop()` from `repair_service::stop()` to `repair_service::shutdown()`.

By convention, shutdown should just signal whatever is running to abort, here, it could happen using the task manager abort_source.
Then, stop() waits for everything to actually stop before destroying the object.

So the repair modude stop can remain being called from repair service stop, given the abort is requested at shutdown time and is being consulted on the the task run path (not to start any new tasks, checked in loops, e.g. per each token range), and used for sleep_abirtable and for waiting e.g. sem.wait or on async condition variables.

> And then we could freely use the `_gate` and maybe even delete `with_gate` from:
> 
> https://github.com/scylladb/scylladb/blob/6b180bf2c7830a5930ffcc87a9e4cf3a25db9092/repair/repair.cc#L474-L485
> 
> 
> Since it is already protected by `task_manager::task` gate.
> One thing that comes to mind here is if task manager api should have been unregistered. Some of the already existing services are unregistered, some are not. I do not see the pattern of choice.
> 
> Also I think that the shutdown should happen in the following order:
> 
> * stop tasks

that would be requesting abort via the abort source. At this point no new tasks should start even while the api is still registered.  They could return an appropriate error at this state.

> * unregister services' apis
> * (unregister task manager api)

that will happen when the task manager service is stopped.

> * close task_manager::modules
> * close task_manager
> * close services
>   Thus, be set in `main` in reversed order.

right

> 
> What do you think?

Cc @xemul